Perfect nested sampling functionality

williamjameshandley opened the pull request:
# Description

A set of scripts for generating perfect nested sampling runs which will prove useful in developing techniques such as importance sampling #132. 

Changes:
1. Implemented a spherical gaussian as found in John Skilling's original paper
2. Implemented a correlated gaussian, which can be used to simulate cosmological likelihoods
3. Moved the wedding cake class into a function in the same framework as the two above

# Checklist:

- [X] I have performed a self-review of my own code
- [X] My code is PEP8 compliant (`flake8 anesthetic tests`)
- [X] My code contains compliant docstrings (`pydocstyle --convention=numpy anesthetic`)
- [X] New and existing unit tests pass locally with my changes (`python -m pytest`)
- [X] I have added tests that prove my fix is effective or that my feature works

lukashergt commented on anesthetic/examples/perfect_ns.py:
```diff
@@ -0,0 +1,197 @@
+        logL = logLike(points)
+        samples.append(NestedSamples(points, logL=logL, logL_birth=logL_birth))
+        logL_birth = logL.copy()
+        r = (points**2).sum(axis=-1, keepdims=True)**0.5
```
What shape does `points` have? Is it `(nlive, ndim)`? Such that we are taking the norm across `ndim` and end up with an array `r` of length `len(r)=nlive`?

williamjameshandley replied:
Yes.

lukashergt commented on anesthetic/examples/utils.py:
```diff
@@ -0,0 +1,53 @@
+
+
+def random_covariance(sigmas):
+    """Draw a randomly oriented covariance matrix with axes length sigmas.
+
+    Parameters
+    ----------
+    sigmas, 1d array like
+        Lengths of the axes of the ellipsoid.
+    """
```
Return value? Shape?

williamjameshandley replied:
Updated in c2decc7

lukashergt reviewed (approved):
This could indeed be useful. Did you intend to add more to this or is this ready for review?

See comments inline.

williamjameshandley commented on anesthetic/examples/perfect_ns.py:
```diff
@@ -0,0 +1,197 @@
+        r = (points**2).sum(axis=-1, keepdims=True)**0.5
+
+    samples = merge_nested_samples(samples)
+    samples.logL
```
No. Resolved in 1a958dd

williamjameshandley commented on anesthetic/examples/utils.py:
```diff
@@ -0,0 +1,53 @@
+        The covariance structure of the ellipsoid. Axes have lengths equal to
+        the square root of the eigenvalues of this matrix.
+    """
+    """Generate samples uniformly from the ellipsoid."""
```
Updated in c2decc7

williamjameshandley commented on anesthetic/examples/__init__.py:
```diff
@@ -0,0 +1 @@
+"""Module for generating nested sampling examples."""
```
Agreed ba4707d.

williamjameshandley commented:
I have added a planck gaussian example (`anesthetic.examples.planck_gaussian`) which produces a gaussian likelihood in 6 cosmological parameters with the correct labels, mean, covariance and normalisation.

One subtlety is that this would not give the same evidence as a full planck run, but we could choose this normalisation a-la-[nuisance marginalised likelihoods](https://arxiv.org/abs/2207.11457).

In addition I corrected a bug which testing this threw up (a factor of 2 in the `d_G`) and improved the efficiency of generating these perfect nested sampling runs (namely updating the bounds on the uniform step, and fixing a bug where previously the bounds were assumed to be 0 and 1 for the elliptical step). These proved essential in making this efficient enough to work with 6 parameters.

It is too inefficient with all nuisance parameters sadly.

williamjameshandley commented:
This will also be useful for removing demo.py in #246.

lukashergt commented on tests/test_examples.py:
```diff
@@ -0,0 +1,149 @@
+    assert_allclose(vol, np.linalg.det(cov)**0.5*volume_n_ball(2), rtol=1e-3)
+
+
+def test_perfect_ns_gaussian():
+    np.random.seed(0)
+    nlive = 500
+    ndims = 10
+    sigma = 0.1
+    R = 1
+
+    samples = gaussian(nlive, ndims, sigma, R)
+
+    assert (samples[:-nlive].nlive == nlive).all()
+
+    mean = samples[np.arange(ndims)].mean()
+    assert_allclose(mean, 0, atol=1e-2)
+    cov = samples[np.arange(ndims)].cov()
+    assert_allclose(cov, np.eye(ndims)*sigma**2, atol=1e-3)
+
+    D_KL = log_volume_n_ball(ndims) - ndims/2*np.log(2*np.pi*np.e*sigma**2)
+    assert_allclose(samples.D_KL(), D_KL, atol=3*samples.D_KL(12).std())
+
+    logZ = ndims/2 * np.log(2*np.pi*sigma**2) - log_volume_n_ball(ndims, R)
+    assert_allclose(samples.logZ(), logZ, atol=3*samples.logZ(12).std())
```
Can we add a test for the correct `d_G` in the perfect Gaussian case, which would allow us to catch any missing factor of 2 in the future?

williamjameshandley replied:
Done in ee1513d

lukashergt commented on anesthetic/examples/utils.py:
```diff
@@ -0,0 +1,63 @@
+    Returns
+    -------
+    Covariance matrix: 2d np.array
+        shape (len(sigmas, len(sigmas)).
```
Missing closing parenthesis.

williamjameshandley replied:
Done in 80d6806

lukashergt commented on anesthetic/examples/utils.py:
```diff
@@ -0,0 +1,63 @@
+    -------
+    points: array-like
+        Points drawn uniformly from the ellipsoid.
+        shape (*size, len(mean)).
```
Why the star?

williamjameshandley replied:
You're right this is neither correct nor clear. I've updated in 62d2c17 the docstring so that it mirrors `np.random.multivariate_normal` which it's supposed to

lukashergt reviewed (commented):
See comments inline.

lukashergt reviewed (approved):
Great! Thanks @williamjameshandley, please squash and merge.
